### PR TITLE
feat: smarter verification

### PR DESCRIPTION
**Pull request by klkvr**
## Motivation

There is no need in requiring contract name/path in `forge verify-contract`. For example, hardhat does not require that.

## Solution

This PR introduces smarter way of identifying contracts by deployed code which respects immutable/link references and targets exact bytecode match (`ContractsByArtifact::find_by_deployed_code_exact`). I believe this could be useful for stuff unrelated to verification as current diff score method is not really great.

Also, this PR makes `contract: ContractInfo` parameter on `VerifyArgs` optional, thus allowing to do just `forge verify-contract <address>`. The artifact to use is determined by matching local bytecode through `find_by_deployed_code_exact`.

I've also changed flow to not rely on cache (except for determining version by looking at recent artifacts). We now trigger compiler directly if we need to get cached abi or metadata and it just uses cache if it's available instead of throwing a error when run on a project without cache.

In combination with `--guess-constructor-args` it now allows users to verify deployments of arbitrary project contracts much easier by just doing `forge verify-contract <address> --guess-constructor-args`

Next steps could probably be identifying libraries used when linking contract automatically as well.

### CODE REVIEW

File: crates/common/src/contracts.rs [hunk 1]
```diff
@@ -1,17 +1,24 @@
...
+    },
     ArtifactId,
 };
 use std::{
     collections::BTreeMap,
     ops::{Deref, DerefMut},
+    str::FromStr,
 };
 
+/// https://docs.soliditylang.org/en/latest/contracts.html#call-protection-for-libraries
```
Comment (DaniPopes) [inline crates/common/src/contracts.rs:19]:
```suggestion
/// Libraries' runtime code always starts with the following instruction:
/// `PUSH20 0x0000000000000000000000000000000000000000`
///
/// See: <https://docs.soliditylang.org/en/latest/contracts.html#call-protection-for-libraries>
```

File: crates/common/src/contracts.rs [hunk 2]
```diff
@@ -1,17 +1,24 @@
...
+const CALL_PROTECTION_BYTECODE_PREFIX: Bytes = bytes!("730000000000000000000000000000000000000000");
```
Comment (DaniPopes) [inline crates/common/src/contracts.rs:20]:
```suggestion
const CALL_PROTECTION_BYTECODE_PREFIX: &[u8] = hex!("730000000000000000000000000000000000000000");
```

File: crates/verify/src/etherscan/flatten.rs [hunk 3]
```diff
@@ -1,10 +1,12 @@
+use crate::provider::VerificationContext;
+
```
Comment (DaniPopes) [inline crates/verify/src/etherscan/flatten.rs:2]:
```suggestion
```

File: crates/verify/src/etherscan/standard_json.rs [hunk 4]
```diff
@@ -1,35 +1,36 @@
+use crate::provider::VerificationContext;
+
```
Comment (DaniPopes) [inline crates/verify/src/etherscan/standard_json.rs:2]:
```suggestion
```

**Review by mattsse — APPROVED** [review]
great, pending @DaniPopes nits

File: crates/common/src/contracts.rs [hunk 5]
```diff
@@ -73,14 +88,107 @@ impl ContractsByArtifact {
...
                 false
             }
         })
     }
 
+    /// Finds a contract which deployed bytecode exactly matches the given code. Accounts for link
+    /// references and immutables.
+    pub fn find_by_deployed_code_exact(&self, code: &[u8]) -> Option<ArtifactWithContractRef> {
+        self.iter().find(|(_, contract)| {
+            let Some(ref deployed_bytecode) = contract.deployed_bytecode else {
```
Comment (DaniPopes) [inline crates/common/src/contracts.rs:103]: pls use & in the expression instead of ref in the pattern where possible

**Review by DaniPopes — APPROVED** [review]

**Comment by mattsse** [comment]
eth_feeHistory bug should go away with #7934

**Pull request merged**